ejb3/src/main/java/org/jboss/as/ejb3/deployment/EJBSecurityDomainService.java

```diff
@@ -36,39 +41,81 @@
 import org.jboss.msc.service.StartException;
 import org.jboss.msc.service.StopContext;
 import org.jboss.msc.value.InjectedValue;
+import org.wildfly.security.auth.server.SecurityDomain;
+import org.wildfly.security.manager.WildFlySecurityManager;
 
 /**
  * A service that sets up the security domain mapping for an EJB deployment.
  *
+ * This service is reponsible for deployment level actions such as registering a deployment with the resolved
+ * application-security-domain mapping or associating the {@code SecurityDomain} with the deployment's {@code ClassLoader}.
+ * Defined {@code ComponentConfigurator} instances will then provide the component specific installation.
+ *
  * @author <a href="mailto:fjuma@redhat.com">Farah</a>
  */
 public class EJBSecurityDomainService implements Service<Void> {
     public static final ServiceName SERVICE_NAME = ServiceName.of("ejb3", "security-domain");
 
     private final InjectedValue<ApplicationSecurityDomain> applicationSecurityDomain = new InjectedValue<>();
+    private final InjectedValue<SecurityDomain> securityDomain = new InjectedValue<>();
+
     private final DeploymentUnit deploymentUnit;
-    private Registration registration;
+
+    private volatile Runnable cleanUpTask;
 
     public EJBSecurityDomainService(final DeploymentUnit deploymentUnit) {
         this.deploymentUnit = deploymentUnit;
     }
 
     @Override
     public synchronized void start(StartContext context) throws StartException {
-        ApplicationSecurityDomain applicationSecurityDomain = getApplicationSecurityDomain();
-        BiFunction<String, ClassLoader, Registration> securityFunction = applicationSecurityDomain != null ? applicationSecurityDomain.getSecurityFunction() : null;
-        if (securityFunction != null) {
-            final String deploymentName = deploymentUnit.getParent() == null ? deploymentUnit.getName() : deploymentUnit.getParent().getName() + "." + deploymentUnit.getName();
-            final Module module = deploymentUnit.getAttachment(Attachments.MODULE);
-            final ClassLoader classLoader = module.getClassLoader();
-            registration = securityFunction.apply(deploymentName, classLoader);
+        final String deploymentName = deploymentUnit.getParent() == null ? deploymentUnit.getName()
+                : deploymentUnit.getParent().getName() + "." + deploymentUnit.getName();
+        final Module module = deploymentUnit.getAttachment(Attachments.MODULE);
+        final ClassLoader classLoader = module.getClassLoader();
+
+        ApplicationSecurityDomain applicationSecurityDomain = this.applicationSecurityDomain.getOptionalValue();
+        if (applicationSecurityDomain != null) {
+            BiFunction<String, ClassLoader, Registration> securityFunction = applicationSecurityDomain != null ? applicationSecurityDomain.getSecurityFunction() : null;
```

Comment:
Nit: you don't need the null check any more as the line above ensures it's false.

Reply:
True, I can clean that - there was an intermediate bug during development where that prior line didn't exist.